pkg/gardenlet/operation/botanist/kubeapiserverexposure.go

```diff
@@ -67,24 +67,25 @@ func (b *Botanist) DeployKubeAPIServerSNI(ctx context.Context) error {
 	return b.Shoot.Components.ControlPlane.KubeAPIServerSNI.Deploy(ctx)
 }
 
-func (b *Botanist) setAPIServerServiceClusterIP(clusterIP string) {
-	clusterIPv4 := net.ParseIP(clusterIP).To4()
+func (b *Botanist) setAPIServerServiceClusterIPs(clusterIPs []string) {
+	clusterIPv4 := net.ParseIP(clusterIPs[0]).To4()
 
 	if clusterIPv4 != nil {
-		if b.Shoot.GetInfo().Spec.Networking.IPFamilies[0] == v1beta1.IPFamilyIPv6 {
+		if b.Shoot.GetInfo().Spec.Networking.IPFamilies[0] == gardencorev1beta1.IPFamilyIPv6 {
 			// "64:ff9b:1::" is a well known prefix for address translation for use
 			// in local networks.
-			b.APIServerClusterIP = "64:ff9b:1::" + clusterIP
+			b.APIServerClusterIP = "64:ff9b:1::" + clusterIPs[0]
 		} else {
-			// prevent leakage of real cluster ip to shoot. we use the reserved range 240.0.0.0/8 as prefix instead.
-			// e.g. cluster ip in seed:  192.168.102.23 => ip in shoot:  240.168.102.23
-			prefixIp, _, _ := net.ParseCIDR(v1beta1constants.ReservedKubeApiServerMappingRange)
-			prefix := prefixIp.To4()
-			b.APIServerClusterIP = net.IPv4(prefix[0], clusterIPv4[1], clusterIPv4[2], clusterIPv4[3]).String()
+			b.APIServerClusterIP = mapToReservedKubeApiServerRange(clusterIPv4)
 		}
 	} else {
-		// regular ipv6 cluster ip
-		b.APIServerClusterIP = clusterIP
+		if gardencorev1beta1.IsIPv4SingleStack(b.Shoot.GetInfo().Spec.Networking.IPFamilies) && len(clusterIPs) > 1 {
```

Comment:
Can you explain under which circumstances a service has `len(clusterIPs) > 1` for IPv4 only clusters?

Shouldn't it be `if !gardencorev1beta1.IsIPv4SingleStack(b.Shoot.GetInfo().Spec.Networking.IPFamilies)`?

Reply:
As discussed, the shoot is IPv4 only while we have a dual stack seed. In this case the IPv4 address needs to be considered to set up the API Server Proxy in the shoot.